pyDeltaRCM/sed_tools.py

```diff
@@ -22,275 +24,585 @@ def sed_route(self):
+                                                       num_starts)
+
+        self._sr.run(start_indices, self.eta, self.stage, self.depth,
+                     self.cell_type, self.uw, self.ux, self.uy,
+                     self.pad_stage, self.pad_depth, self.pad_cell_type,
+                     self.Vp_dep_mud, self.qw, self.qx, self.qy, self.qs)
+
+        self.Vp_dep_mud = self._sr.Vp_dep_mud
+        self.eta = self._sr.eta  # update bed
+        self.depth = self._sr.depth  # update depth
+        self.pad_depth = self._sr.pad_depth  # update depth in padded array
+        self.uw = self._sr.uw  # update absolute flow field
+        self.ux = self._sr.ux  # update component flow field
+        self.uy = self._sr.uy  # update component flow fielda
+        self.qs = self._sr.qs
+
+    def route_all_mud_parcels(self):
+        """Route mud parcels."""
+        num_starts = int(self.Np_sed * (1 - self.f_bedload))
+        inlet_weights = np.ones_like(self.inlet)
+        start_indices = shared_tools.get_start_indices(self.inlet,
+                                                       inlet_weights,
+                                                       num_starts)
+
+        self._mr.run(start_indices, self.eta, self.stage, self.depth,
+                     self.cell_type, self.uw, self.ux, self.uy,
+                     self.pad_stage, self.pad_depth, self.pad_cell_type,
+                     self.Vp_dep_mud, self.qw, self.qx, self.qy)
+
+        self.Vp_dep_mud = self._mr.Vp_dep_mud
+        self.eta = self._mr.eta  # update bed
+        self.depth = self._mr.depth  # update depth
+        self.pad_depth = self._mr.pad_depth  # update depth in padded array
+        self.uw = self._mr.uw  # update absolute flow field
+        self.ux = self._mr.ux  # update component flow field
+        self.uy = self._mr.uy  # update component flow field
 
-        Returns
-        -------
+    def topo_diffusion(self):
+        """Diffuse topography after routing.
 
+        Diffuse topography after routing all coarse sediment parcels. The
+        method uses convolution with a kernel to compute smoother topography,
+        and then adds this different to the current eta to do the smoothing.
+        The operation is repeated `N_crossdiff` times.
         """
-        eta_change_loc = Vp_dep / self.dx**2
-
-        self.eta[px, py] = self.eta[px, py] + eta_change_loc  # update bed
-        self.depth[px, py] = self.stage[px, py] - self.eta[px, py]
-
-        if self.depth[px, py] < 0:
-            self.depth[px, py] = 0
+        for _ in range(self.N_crossdiff):
 
-        self.pad_depth[px + 1, py + 1] = self.depth[px, py]
+            a = ndimage.convolve(self.eta, self.kernel1, mode='constant')
+            b = ndimage.convolve(self.qs, self.kernel2, mode='constant')
+            c = ndimage.convolve(self.qs * self.eta, self.kernel2,
+                                 mode='constant')
```

Comment:
This is a very cool way of accomplishing this.

Reply:
this predates my involvement the project, so I'm not sure who implemented it this way.